File: js/problem39.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */

const getCandidates = function(candidates, target, start, currSum, curr, res) {
    if (currSum >= target) {
        if (currSum === target) {
            res.push(curr.slice());
        }
        return;
    }
    
    for (let i = start; i < candidates.length; i++) {
        curr.push(candidates[i]);
        getCandidates(candidates, target, i, currSum + candidates[i], curr, res);
        curr.pop();
    }
};


var combinationSum = function(candidates, target) {
    if (candidates.length === 0) {
        return [];
    }
    
    let res = [];
    
    for (let i = 0; i < candidates.length; i++) {
        let curr = [candidates[i]];
        getCandidates(candidates, target, i, candidates[i], curr, res);
    }
    
    return res;
};

File: leetcodeJava/src/main/java/Problem39.java
// 39. Combination Sum

class Solution {
    
    private void backtrack(int[] candidates, int target, int current, int currentSum, Stack<Integer> stack, List<List<Integer>> res) {
        
        if (currentSum >= target) {
            if (currentSum == target) {
                List<Integer> temp = new ArrayList<Integer>();
                for (int n : stack) {
                    temp.add(n);
                }
                res.add(temp);
            }
            return;
        }
        
        for (int i = current; i < candidates.length; i++) {
            stack.push(candidates[i]);
            this.backtrack(candidates, target, i, currentSum + candidates[i], stack, res);
            stack.pop();
        }
    }
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        Stack<Integer> stack = new Stack<Integer>();
        
        for (int i = 0; i < candidates.length; i++) {
            stack.push(candidates[i]);
            this.backtrack(candidates, target, i, candidates[i], stack, res);
            stack.pop();
        }
        
        return res;
    }
}